File: algospot/boardcover2.py
# f = open("boardcover2.in")
# input = f.readline
# input = raw_input


def rotate(block):
    return zip(*reversed(block))


def relativeize(block):
    basei = basej = -1
    ret = []
    for i, row in enumerate(block):
        for j, cell in enumerate(row):
            if cell == 1:
                if basei == -1:
                    basei, basej = i, j
                ret.append((i-basei, j-basej))
    return ret


def place_block(i, j, rot, add):
    bSuc = True
    for di, dj in rot:
        if 0 <= di+i < R and 0 <= dj+j < C:
            board[di+i][dj+j] += add
            if board[di+i][dj+j] > 1:
                bSuc = False
        else:
            bSuc = False
            break
    return bSuc


def search(placed, blanks):
    global best
    if blanks / blocksize + placed <= best:
        return

    for i in range(R):
        for j in range(C):
            if board[i][j] == 0:
                break
        else:
            continue
        break
    else:
        best = max(best, placed)
        return

    for rot in rotations:
        if place_block(i, j, rot, 1):
            search(placed+1, blanks - blocksize)
        place_block(i, j, rot, -1)

    board[i][j] = 1
    search(placed, blanks - 1)
    board[i][j] = 0


def binarize(cell):
    return 1 if cell == '#' else 0

for case in range(int(input())):
    R, C, BR, BC = map(int, input().split())
    board = [list(map(binarize, input())) for r in range(R)]
    block = [list(map(binarize, input())) for r in range(BR)]

    rotations = []
    for r in range(4):
        rot = relativeize(block)
        if rot not in rotations:
            rotations.append(rot)
        block = list(rotate(block))

    best = 0
    blanks = R*C - sum(sum(line) for line in board)
    blocksize = len(rotations[0])

    search(0, blanks)
    print(best)

File: algospot/boardcover2.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int h, w, bh, bw, blockSize, best, blanks;
int covered[10][10];
vector<vector<pair<int,int>>> rotations;

vector<string> rotate(const vector<string>& arr) {
	vector<string> ret(arr[0].size(), string(arr.size(), ' '));
	for (int i = 0; i < arr.size(); i++)
		for (int j = 0; j < arr[0].size(); j++)
			ret[j][arr.size()-i-1] = arr[i][j];
	return ret;
}

void generateRotations(vector<string> block)  {
	rotations.clear();
	rotations.resize(4);
	for (int rot = 0; rot < 4; rot++) {
		int oy = -1, ox = -1;
		for (int i = 0; i < block.size(); i++) {
			for (int j = 0; j < block[i].size(); j++) {
				if (block[i][j] == '#') {
					if (oy == -1) {
						oy = i;
						ox = j;
					}
					rotations[rot].push_back({i-oy, j-ox});
				}
			}
		}
		block = rotate(block);
	}
	sort(rotations.begin(), rotations.end());
	rotations.erase(unique(rotations.begin(), rotations.end()), rotations.end());
	blockSize = rotations[0].size();
}

bool set(int y, int x, const vector<pair<int,int>>& block, int delta) {
	auto bRet = true;
	for (auto& b : block) {
		auto yy = y + b.first;
		auto xx = x + b.second;
		if (xx < 0 || xx >= w || yy < 0 || yy >= h) {
			return false;
		} else {
			covered[yy][xx] += delta;
			if (covered[yy][xx] > 1)
				bRet = false;
		}
	}
	return bRet;
}

void search(int placed, int marked) {
	auto used = placed*blockSize + marked;
	if ((blanks - used)/blockSize + placed <= best)
		return;
	int y = -1, x = -1;
	for (int i = 0; i < h; i++) {
		for (int j = 0; j < w; j++) {
			if (covered[i][j] == 0) {
				y = i;
				x = j;
				break;
			}
		}
		if (y != -1) 
			break;
	}
	if (y == -1) {
		best = max(placed, best);
		return;
	}
	for (int i = 0; i < rotations.size(); i++) {
		if (set(y, x, rotations[i], 1))
			search(placed+1, marked);
		set(y, x, rotations[i], -1);	
	}
	covered[y][x] = 1;
	search(placed, marked+1);
	covered[y][x] = 0;
}

int main() {
#ifdef _DEBUG
	freopen("boardcover2.in", "r", stdin);
#endif
	int c;
	string str;
	cin >> c;
	while (c--) {
		cin >> h >> w >> bh >> bw;
		vector<string> board(h);
		for (int i = 0; i < h; i++) 
			cin >> board[i];
		vector<string> block(bh);
		for (int i = 0; i < bh; i++) 
			cin >> block[i];
		generateRotations(block);
		best = 0, blanks = 0;
		for (int i = 0; i < h; i++) {
			for (int j = 0; j < w; j++) { 
				covered[i][j] = (board[i][j] == '#' ? 1 : 0);
				blanks += (covered[i][j] == 0);
			}
		}
		search(0, 0);
		cout << best << endl;
	}
	return 0;
}
